text/0000-export-executable-symbols.md

```diff
@@ -0,0 +1,116 @@
+[LINK](https://github.com/MaulingMonkey/jerk/blob/04250c9d1b6ccc292eb27663f70919345c31007f/example-hello-world-jar/build.rs#L4).
+This is ugly, brittle, and rustc
+[already knows](https://github.com/rust-lang/rust/blob/a916ac22b9f7f1f0f7aba0a41a789b3ecd765018/src/librustc_codegen_ssa/back/linker.rs#L706-L717)
+how to do this automatically, across more platforms, and better.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+https://doc.rust-lang.org/rustc/codegen-options/index.html could gain:
+
+```md
+## export-executable-symbols
+
+This flag causes `rustc` to export symbols from executables, as if they were dynamic libraries.
+
+You might use this to allow the JVM or MSCLR to call back into your executable's
+Rust code from Java/C# when embedding their runtimes into your Rust executable.
+```
+
+`rustc -C help` could gain:
+
+```
+    -C    export-executable-symbols -- export symbols from executables, as if they were dynamic libraries.
+```
+
+My Java interop [Quick Start](https://github.com/MaulingMonkey/jerk/blob/master/Readme.md#quick-start)
+would start recommending a `.cargo/config` with:
+```toml
+[build]
+rustflags = ["-C", "export-executable-symbols"]
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+On a technical level, this just involves preventing an early bailout when
+calling `fn export_symbols` on executables with MSVC or GNU linker backends.
+Other linker backends (EmLinker, WasmLd, PtxLinker) do not have this early
+bailout in the first place, and remain unaffected.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* Options bloat
+* The burden of supporting a niche use-case in hideously platform specific code
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This is *very* simple to implement, leverages existing code to enable it to do exactly what it was meant to do, and has few drawbacks.
+
+Alternatives:
+
+- Unconditionally export symbols from executables instead of introducing a new compiler flag.
+- Introduce a crate-level attribute instead of a compiler flag (`#![export_all_symbols]`? `#![export_symbols]`?)
+- Write *yet another* cargo subcommand to install/remember for interop testing instead of using cargo test.
+- Write interop tests exclusively as integration tests, in an entirely separate crate, that can load the testee as a dylib.
+- Continue abusing LINK, writing a tool to auto-generate .defs via build scripts - possibly by reading metadata from other tools.
+- Use nightly link-args instead of LINK, but still write a .def generator.
+- Remember to always cargo build a dylib copy of a crate manually before cargo test ing, and load that instead.
+  (That would also add a whole second copy of all functions and static vars in the same unit test process!)
+
+# Prior art
+[prior-art]: #prior-art
+
+C and C++ compilers can already do this via `__declspec(dllexport)` annotations.
```

Comment:
dllexport was a part of a previous (closed) RFC https://github.com/rust-lang/rfcs/pull/1296 and discussed fairly extensively. Might be worth it to look through that RFC to see if we didn’t miss anything.